lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -855,6 +855,126 @@ static void optimizeBatchNorm(Function *F) {
+        continue;
+      }
+      if (inputs[0].dims()[j] != inputs[i].dims()[j]) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+// Given one tensor and the desired sub-tensor size, \returns the dimension,
+// after which the sub-tensor has the desired size, otherwise return -1.
+static ssize_t findMatchingConcatDimForSameSizeSubTensor(
+    llvm::ArrayRef<unsigned long> firstDims, size_t desiredSubTensorSize) {
+  size_t currentSubTensorSize = 1;
+  ssize_t dimension = -1;
+  for (ssize_t i = firstDims.size() - 1; i >= 0; i--) {
+    if (currentSubTensorSize == desiredSubTensorSize) {
+      dimension = i;
+      break;
+    }
+    currentSubTensorSize *= firstDims[i];
+  }
+  return dimension;
+}
+
+/// Given input tensors and a original ConcatNode, try to find out if
+/// there is a dimension in the input tensors, with which we can meet
+/// two requirements:
+/// 1) Input tensors are concatenate-able along this dimension.
+/// 2) The sub-tensors after this dimension in the input tensors, are
+///    of the same size as the sub-tensor of the input of the original
+///    Concat node after the concatenation dimension.
+/// \returns this dimension if found, otherwise -1.
+static int tryToConcatAlongSameSizeSubTensor(
+    const llvm::SmallVectorImpl<NodeValue> &inputs, ConcatNode *origConcatN) {
+  // For the purpose of the optimiztion
+  // Concat(Reshape(X)*N)->Reshape(Concat(N*X)), we want to make sure the new
+  // ConcatNode can concatenate on the sub-tensor which is of the same size of
+  // that of the original Concate node.
+
+  auto firstDims = inputs.front().dims();
+  auto origCNInputDims = origConcatN->getInputs().front().dims();
+  // The size of the sub-tensors of the original ConcatNode, which are being
+  // concatenated. This size is simply the product of dimensions following the
+  // dimension used for concatenation.
+  int subTensorSizeOriginalConcatNode = 1;
+  for (size_t i = origConcatN->getDim() + 1; i < origCNInputDims.size(); ++i) {
+    subTensorSizeOriginalConcatNode *= origCNInputDims[i];
+  }
+
+  // Try to find the dimension in the first input such that the sub-tensor size
+  // based on this direction is the same as the size of sub-tensors based on the
+  // concatenation dimension used by the original ConcatNode.
+  ssize_t dimension = findMatchingConcatDimForSameSizeSubTensor(
+      firstDims, subTensorSizeOriginalConcatNode);
+  if (dimension == -1) {
+    return -1;
+  }
+
+  // Now we have found the dimension, we need to check if all inputs can be
+  // concatenated along this dimension.
+  if (!verifyInputsForConcat(inputs, dimension)) {
+    return -1;
+  }
+  return dimension;
+}
+
+// Given the inputs of one Concat Nodes, \returns true if they are all
+// ReshapeNode, and the input tensors of these input nodes have same number of
+// dimensions, otherwise returns false.
+static bool
+tryToGetNewConcatInputs(const llvm::ArrayRef<NodeValue> originalConcatInputs,
```

Comment:
ArrayRef does not need to be 'const'. It's already const.

Reply:
thx for the review and explanation!